src/java/org/apache/nutch/indexer/IndexerMapReduce.java

```diff
@@ -238,38 +238,37 @@ public void reduce(Text key, Iterator<NutchWritable> values,
     }
 
     // Whether to delete GONE or REDIRECTS
-    if (delete && fetchDatum != null && dbDatum != null) {
-      if (fetchDatum.getStatus() == CrawlDatum.STATUS_FETCH_GONE
-          || dbDatum.getStatus() == CrawlDatum.STATUS_DB_GONE) {
+    if (delete) {
```

Comment:
I would use
```
if (delete && fetchDatum != null) {
```
and skip further null checks of fetchDatum. In a continuous crawl the CrawlDb will hold many redirects and gone pages, significantly more than in the current segment which is indexed. Sending index deletions for all of them will slow down the index step.

Reply:
I will make the change, but isn't this code running only on the specific segments (and URLs) that were Fetched in this cycle?

Reply:
Yes, but the indexer takes also the full CrawlDb (and optionally LinkDb) as input. In a crawl with many cycles the CrawlDb may become significantly bigger than the currently processed segment(s). There is no way to filter CrawlDb items without counter-parts in the segments beforehand, all end up in the reducer.